File: web/src/api/role.ts
/**
 * 角色管理 API
 */

import { get, post, put, del } from '@/utils/request'

// ==================== 类型定义 ====================

/**
 * 角色信息
 */
export interface Role {
  id: number
  name: string
  code: string
  description: string
  is_system: boolean // 是否为系统角色（不可删除）
  created_at: string
  updated_at: string
  permissions?: RolePermission[] // 角色权限列表（可选）
}

/**
 * 角色权限（按资源类型分组）
 */
export interface RolePermission {
  id: number
  role_id: number
  resource_type: string // 资源类型（如 "directory", "table", "form", "chart"）
  action: string // 权限点（如 "directory:read", "function:write"）
}

/**
 * 创建角色请求
 */
export interface CreateRoleReq {
  name: string
  code: string
  description?: string
  permissions: Record<string, string[]> // 权限点列表，按资源类型分组，例如：{ "directory": ["directory:read", "directory:write"], "table": ["function:read", "function:write"] }
}

/**
 * 创建角色响应
 */
export interface CreateRoleResp {
  role: Role
}

/**
 * 更新角色请求
 */
export interface UpdateRoleReq {
  name?: string
  description?: string
  permissions?: Record<string, string[]> // 权限点列表，按资源类型分组
}

/**
 * 更新角色响应
 */
export interface UpdateRoleResp {
  role: Role
}

/**
 * 获取角色响应
 */
export interface GetRoleResp {
  role: Role
}

/**
 * 获取角色列表响应
 */
export interface GetRolesResp {
  roles: Role[]
}

/**
 * 删除角色响应
 */
export interface DeleteRoleResp {
  message: string
}

// ==================== 角色分配相关 ====================

/**
 * 给用户分配角色请求
 */
export interface AssignRoleToUserReq {
  user: string // 工作空间所属用户
  app: string // 工作空间应用代码
  username: string // 要分配角色的用户名
  role_code: string // 角色代码
  resource_path: string // 资源路径（支持通配符，如 "/user/app/*"）
  start_time?: string // 开始时间（ISO 8601 格式，可选）
  end_time?: string // 结束时间（ISO 8601 格式，可选）
}

/**
 * 给用户分配角色响应
 */
export interface AssignRoleToUserResp {
  assignment: RoleAssignment
}

/**
 * 给组织架构分配角色请求
 */
export interface AssignRoleToDepartmentReq {
  user: string // 工作空间所属用户
  app: string // 工作空间应用代码
  department_path: string // 组织架构路径（如 "/org/master/bizit"）
  role_code: string // 角色代码
  resource_path: string // 资源路径（支持通配符，如 "/user/app/*"）
  start_time?: string // 开始时间（ISO 8601 格式，可选）
  end_time?: string // 结束时间（ISO 8601 格式，可选）
}

/**
 * 给组织架构分配角色响应
 */
export interface AssignRoleToDepartmentResp {
  assignment: RoleAssignment
}

/**
 * 角色分配信息
 */
export interface RoleAssignment {
  id: number
  user: string
  app: string
  subject_type: 'user' | 'department'
  subject: string // 用户名或组织架构路径
  role_code: string
  resource_path: string
  start_time?: string
  end_time?: string
  created_at: string
  updated_at: string
}

/**
 * 移除用户角色请求
 */
export interface RemoveRoleFromUserReq {
  user: string
  app: string
  username: string
  role_code: string
  resource_path: string
}

/**
 * 移除用户角色响应
 */
export interface RemoveRoleFromUserResp {
  message: string
}

/**
 * 移除组织架构角色请求
 */
export interface RemoveRoleFromDepartmentReq {
  user: string
  app: string
  department_path: string
  role_code: string
  resource_path: string
}

/**
 * 移除组织架构角色响应
 */
export interface RemoveRoleFromDepartmentResp {
  message: string
}

/**
 * 获取用户角色请求
 */
export interface GetUserRolesReq {
  user: string
  app: string
  username: string
}

/**
 * 获取用户角色响应
 */
export interface GetUserRolesResp {
  assignments: RoleAssignment[]
}

/**
 * 获取组织架构角色请求
 */
export interface GetDepartmentRolesReq {
  user: string
  app: string
  department_path: string
}

/**
 * 获取组织架构角色响应
 */
export interface GetDepartmentRolesResp {
  assignments: RoleAssignment[]
}

// ==================== API 调用 ====================

/**
 * 获取所有角色
 * @param resourceType 可选的资源类型过滤（directory、table、form、chart、app）
 */
export function getRoles(resourceType?: string): Promise<GetRolesResp> {
  const params = resourceType ? { resource_type: resourceType } : undefined
  return get<GetRolesResp>('/workspace/api/v1/role', params)
}

/**
 * 获取角色详情
 */
export function getRole(id: number): Promise<GetRoleResp> {
  return get<GetRoleResp>(`/workspace/api/v1/role/${id}`)
}

/**
 * 创建角色
 */
export function createRole(data: CreateRoleReq): Promise<CreateRoleResp> {
  return post<CreateRoleResp>('/workspace/api/v1/role', data)
}

/**
 * 更新角色
 */
export function updateRole(id: number, data: UpdateRoleReq): Promise<UpdateRoleResp> {
  return put<UpdateRoleResp>(`/workspace/api/v1/role/${id}`, data)
}

/**
 * 删除角色
 */
export function deleteRole(id: number): Promise<DeleteRoleResp> {
  return del<DeleteRoleResp>(`/workspace/api/v1/role/${id}`)
}

/**
 * 给用户分配角色
 */
export function assignRoleToUser(data: AssignRoleToUserReq): Promise<AssignRoleToUserResp> {
  return post<AssignRoleToUserResp>('/workspace/api/v1/role/assign/user', data)
}

/**
 * 给组织架构分配角色
 */
export function assignRoleToDepartment(data: AssignRoleToDepartmentReq): Promise<AssignRoleToDepartmentResp> {
  return post<AssignRoleToDepartmentResp>('/workspace/api/v1/role/assign/department', data)
}

/**
 * 移除用户角色
 */
export function removeRoleFromUser(data: RemoveRoleFromUserReq): Promise<RemoveRoleFromUserResp> {
  return post<RemoveRoleFromUserResp>('/workspace/api/v1/role/remove/user', data)
}

/**
 * 移除组织架构角色
 */
export function removeRoleFromDepartment(data: RemoveRoleFromDepartmentReq): Promise<RemoveRoleFromDepartmentResp> {
  return post<RemoveRoleFromDepartmentResp>('/workspace/api/v1/role/remove/department', data)
}

/**
 * 获取用户角色
 */
export function getUserRoles(data: GetUserRolesReq): Promise<GetUserRolesResp> {
  return post<GetUserRolesResp>('/workspace/api/v1/role/user', data)
}

/**
 * 获取组织架构角色
 */
export function getDepartmentRoles(data: GetDepartmentRolesReq): Promise<GetDepartmentRolesResp> {
  return post<GetDepartmentRolesResp>('/workspace/api/v1/role/department', data)
}

/**
 * 获取可用于权限申请的角色列表请求
 */
export interface GetRolesForPermissionRequestReq {
  node_type: string // 节点类型 (package, function)
  template_type?: string // 模板类型 (table, form, chart)
}

/**
 * 获取可用于权限申请的角色列表响应
 */
export interface GetRolesForPermissionRequestResp {
  roles: Role[]
}

/**
 * 获取可用于权限申请的角色列表（根据节点类型过滤）
 */
export function getRolesForPermissionRequest(params: GetRolesForPermissionRequestReq): Promise<GetRolesForPermissionRequestResp> {
  return get<GetRolesForPermissionRequestResp>('/workspace/api/v1/role/for_request', params)
}

File: dto/role.go
package dto

import (
	"github.com/ai-agent-os/ai-agent-os/core/app-server/model"
	"github.com/ai-agent-os/ai-agent-os/pkg/gormx/models"
)

// ============================================
// 角色管理 DTO
// ============================================

// CreateRoleReq 创建角色请求
type CreateRoleReq struct {
	Name        string              `json:"name" binding:"required" example:"开发者"`
	Code        string              `json:"code" binding:"required" example:"developer"`
	Description string              `json:"description" example:"开发人员角色"`
	Permissions map[string][]string `json:"permissions" binding:"required" example:"{\"directory\":[\"directory:read\",\"directory:write\"],\"table\":[\"function:read\",\"function:write\"]}"` // resourceType -> []action
}

// CreateRoleResp 创建角色响应
type CreateRoleResp struct {
	Role *model.Role `json:"role"`
}

// UpdateRoleReq 更新角色请求
type UpdateRoleReq struct {
	Name        *string              `json:"name,omitempty" example:"开发者"`
	Description *string              `json:"description,omitempty" example:"开发人员角色"`
	IsDefault   *bool                `json:"is_default,omitempty" example:"true"` // 是否设置为默认角色
	Permissions *map[string][]string `json:"permissions,omitempty" example:"{\"directory\":[\"directory:read\"],\"table\":[\"function:read\"]}"` // resourceType -> []action
}

// UpdateRoleResp 更新角色响应
type UpdateRoleResp struct {
	Role *model.Role `json:"role"`
}

// GetRoleResp 获取角色响应
type GetRoleResp struct {
	Role *model.Role `json:"role"`
}

// GetRolesResp 获取角色列表响应
type GetRolesResp struct {
	Roles []*model.Role `json:"roles"`
}

// DeleteRoleResp 删除角色响应
type DeleteRoleResp struct {
	Message string `json:"message"`
}

// ============================================
// 角色分配 DTO
// ============================================

// AssignRoleToUserReq 给用户分配角色请求
type AssignRoleToUserReq struct {
	User         string       `json:"user" binding:"required" example:"luobei"`
	App          string       `json:"app" binding:"required" example:"operations"`
	Username     string       `json:"username" binding:"required" example:"zhangsan"`
	RoleCode     string       `json:"role_code" binding:"required" example:"developer"`
	ResourceType string       `json:"resource_type" binding:"required" example:"table"` // ⭐ 资源类型：directory、table、form、chart、app
	ResourcePath string       `json:"resource_path" binding:"required" example:"/luobei/operations/tools/*"`
	StartTime    *models.Time `json:"start_time,omitempty" example:"2025-01-27T00:00:00Z"`
	EndTime      *models.Time `json:"end_time,omitempty" example:"2025-12-31T23:59:59Z"`
}

// AssignRoleToUserResp 给用户分配角色响应
type AssignRoleToUserResp struct {
	Assignment *model.RoleAssignment `json:"assignment"`
}

// AssignRoleToDepartmentReq 给组织架构分配角色请求
type AssignRoleToDepartmentReq struct {
	User           string       `json:"user" binding:"required" example:"luobei"`
	App            string       `json:"app" binding:"required" example:"operations"`
	DepartmentPath string       `json:"department_path" binding:"required" example:"/org/master/bizit"`
	RoleCode       string       `json:"role_code" binding:"required" example:"viewer"`
	ResourceType   string       `json:"resource_type" binding:"required" example:"table"` // ⭐ 资源类型：directory、table、form、chart、app
	ResourcePath   string       `json:"resource_path" binding:"required" example:"/luobei/operations/*"`
	StartTime      *models.Time `json:"start_time,omitempty"`
	EndTime        *models.Time `json:"end_time,omitempty"`
}

// AssignRoleToDepartmentResp 给组织架构分配角色响应
type AssignRoleToDepartmentResp struct {
	Assignment *model.RoleAssignment `json:"assignment"`
}

// RemoveRoleFromUserReq 移除用户角色请求
type RemoveRoleFromUserReq struct {
	User         string `json:"user" binding:"required"`
	App          string `json:"app" binding:"required"`
	Username     string `json:"username" binding:"required"`
	RoleCode     string `json:"role_code" binding:"required"`
	ResourceType string `json:"resource_type" binding:"required"` // ⭐ 资源类型：directory、table、form、chart、app
	ResourcePath string `json:"resource_path" binding:"required"`
}

// RemoveRoleFromUserResp 移除用户角色响应
type RemoveRoleFromUserResp struct {
	Message string `json:"message"`
}

// RemoveRoleFromDepartmentReq 移除组织架构角色请求
type RemoveRoleFromDepartmentReq struct {
	User           string `json:"user" binding:"required"`
	App            string `json:"app" binding:"required"`
	DepartmentPath string `json:"department_path" binding:"required"`
	RoleCode       string `json:"role_code" binding:"required"`
	ResourceType   string `json:"resource_type" binding:"required"` // ⭐ 资源类型：directory、table、form、chart、app
	ResourcePath   string `json:"resource_path" binding:"required"`
}

// RemoveRoleFromDepartmentResp 移除组织架构角色响应
type RemoveRoleFromDepartmentResp struct {
	Message string `json:"message"`
}

// GetUserRolesReq 获取用户角色请求
type GetUserRolesReq struct {
	User     string `json:"user" binding:"required"`
	App      string `json:"app" binding:"required"`
	Username string `json:"username" binding:"required"`
}

// GetUserRolesResp 获取用户角色响应
type GetUserRolesResp struct {
	Assignments []*model.RoleAssignment `json:"assignments"`
}

// GetDepartmentRolesReq 获取组织架构角色请求
type GetDepartmentRolesReq struct {
	User           string `json:"user" binding:"required"`
	App            string `json:"app" binding:"required"`
	DepartmentPath string `json:"department_path" binding:"required"`
}

// GetDepartmentRolesResp 获取组织架构角色响应
type GetDepartmentRolesResp struct {
	Assignments []*model.RoleAssignment `json:"assignments"`
}

// GetRolesForPermissionRequestReq 获取可用于权限申请的角色列表请求
type GetRolesForPermissionRequestReq struct {
	NodeType     string `form:"node_type" json:"node_type" example:"function"`       // 节点类型：package 或 function
	TemplateType string `form:"template_type" json:"template_type" example:"table"` // 模板类型：table、form、chart（仅对 function 有效）
}

// GetRolesForPermissionRequestResp 获取可用于权限申请的角色列表响应
type GetRolesForPermissionRequestResp struct {
	Roles []*model.Role `json:"roles"` // 角色列表（只包含对该资源类型有权限的角色）
}

// ============================================
// 资源权限管理 DTO
// ============================================

// GetResourcePermissionsReq 查询资源权限请求
type GetResourcePermissionsReq struct {
	User         string `json:"user" form:"user" binding:"required" example:"luobei"`         // 租户用户
	App          string `json:"app" form:"app" binding:"required" example:"operations"`       // 应用代码
	ResourcePath string `json:"resource_path" form:"resource_path" binding:"required" example:"/luobei/operations/tools/crm_ticket"` // 资源路径（full-code-path）
}

// ResourcePermissionAssignment 资源权限分配信息
type ResourcePermissionAssignment struct {
	ID           int64     `json:"id" example:"1"`              // 分配ID
	SubjectType  string    `json:"subject_type" example:"user"` // 权限主体类型：user 或 department
	Subject      string    `json:"subject" example:"zhangsan"`   // 权限主体：用户名或组织架构路径
	SubjectName  string    `json:"subject_name" example:"张三"`  // 权限主体名称（用户昵称或部门名称）
	RoleID       int64     `json:"role_id" example:"1"`         // 角色ID
	RoleCode     string    `json:"role_code" example:"developer"` // 角色编码
	RoleName     string    `json:"role_name" example:"开发者"`    // 角色名称
	ResourcePath string       `json:"resource_path" example:"/luobei/operations/tools/crm_ticket"` // 资源路径
	ResourceName string       `json:"resource_name" example:"工单管理"` // 资源名称（节点名称）
	StartTime    models.Time  `json:"start_time" example:"2025-01-27T00:00:00Z"` // 生效开始时间
	EndTime      *models.Time `json:"end_time,omitempty" example:"2025-12-31T23:59:59Z"` // 生效结束时间（nil 表示永久）
	CreatedBy    string       `json:"created_by,omitempty" example:"admin"` // 创建者
	CreatedAt    models.Time  `json:"created_at" example:"2025-01-27T00:00:00Z"` // 创建时间
}

// GetResourcePermissionsResp 查询资源权限响应
type GetResourcePermissionsResp struct {
	Assignments []*ResourcePermissionAssignment `json:"assignments"` // 权限分配列表
	Total       int                              `json:"total" example:"10"` // 总数
}
